test/TPP/linalg-to-tpp.mlir

```diff
@@ -99,3 +99,60 @@ func.func @main() -> memref<8x32x32x32xf32> {
   }
   return %alloc : memref<8x32x32x32xf32>
 }
+
+// -----
+
+#map = affine_map<(d0) -> (d0)>
+func.func @add_mapping(%arg0: memref<1xf32>, %arg1: memref<1xf32>) {
+  // CHECK: tpp.add
+  linalg.generic {indexing_maps = [#map, #map], iterator_types = ["parallel"]} ins(%arg0: memref<1xf32>) outs(%arg1: memref<1xf32>) {
+    ^bb0(%in: f32, %out: f32):
+      %0 = arith.addf %in, %out : f32
+      linalg.yield %0 : f32
+  }
+  return 
+}
+
+// -----
+
+// Scalar operands we don't expect any mapping to tpp.
+#map = affine_map<() -> ()>
+func.func @add_mapping(%arg0: memref<f32>, %arg1: memref<f32>) {
+  // CHECK-NOT: tpp.add
+  linalg.generic {indexing_maps = [#map, #map], iterator_types = []} ins(%arg0: memref<f32>) outs(%arg1: memref<f32>) {
+    ^bb0(%in: f32, %out: f32):
+      %0 = arith.addf %in, %out : f32
+      linalg.yield %0 : f32
+  }
+  return
+}
+
+// -----
+
+#map = affine_map<(d0, d1, d2) -> (d0, d1, d2)>
+func.func @add_mapping(%arg0: memref<10x10x10xf32>, %arg1: memref<10x10x10xf32>) {
+  // CHECK-DAG: %[[C0:.+]] = arith.constant 0 : index
+  // CHECK-DAG: %[[C10:.+]] = arith.constant 10 : index
+  // CHECK-DAG: %[[C1:.+]] = arith.constant 1 : index
+  // CHECK: scf.parallel ({{.*}}) to (%[[C10]]) step (%[[C1]])
+  // CHECK: tpp.add
```

Comment:
We can peel iterations and map to tpp (reshape2D does that).

Reply:
Right, so this is not testing the exit condition in the code. Can you add another test that does that? (and explain in a comment why this test doesn't meet that criteria?)

Reply:
the exit condition is already tested here:

```
// Scalar operands we don't expect any mapping to tpp.
#map = affine_map<() -> ()>
func.func @add_mapping(%arg0: memref<f32>, %arg1: memref<f32>) {
  // CHECK-NOT: tpp.add
  linalg.generic {indexing_maps = [#map, #map], iterator_types = []} ins(%arg0: memref<f32>) outs(%arg1: memref<f32>) {
    ^bb0(%in: f32, %out: f32):
      %0 = arith.addf %in, %out : f32
      linalg.yield %0 : f32
  }
  return
}```

Reply:
Do you want a test that checks the affine maps?

Reply:
Only one condition is checked there, the one where the rank is zero. The other condition, where the rank (minus the 1 dms) is greater than two, isn't.

I'm expecting to see an error when you try to apply the transform on a tensor that is `10x10x10`. If you do other transforms on the existing test, then you're not testing that condition. 

To e clear, we need to have a test, with `CHECK-NOT: tpp.add` because the tensor rank doesn't match the conditions you are creating.

It would also be nice to have tests beyond `1x10x10`, like `10x1x10` and `10x10x1`, as well as `1x1` and `1x1x1`, to make sure there are no corner cases that you missed.

Reply:
This function is part of a conversion pass and we can map tensors like 10x10x10 by pulling out the outermost dimension. Looking at the conversion again there is not need to check rank in `isTppAdd` but rank is ensured to be 2d by `rankReducingSubviewDroppingUnitDims` in `ConvertLinalgToTpp.cpp`. Here we can check only the affine maps. I added more tests as suggested.